jest/jestSetup.js

```diff
@@ -67,6 +67,8 @@ jest.mock('react-native', () => {
   return ReactNative;
 });
 
+jest.mock('immediate', () => cb => Promise.resolve().then(cb));
```

Comment:
It's present but mocked out:
https://jestjs.io/docs/timer-mocks
(see also our async-test.js), and as a result when something running under a test creates a timer (as with `setTimeout`), the timer callback may not be run unless the test calls `jest.runAllTimers()` or one of its friends. Plus, as you can see in async-test.js, and more so in the "our promisified sqlite > transaction with internal asynchrony …" test added in this series, even doing that can get hairy because the `jest.runAllTimers()` isn't effective before the timer has actually been created.

As that same "transaction with internal asynchrony" test illustrates, this applies even if the `setTimeout` duration is zero, as it is in the `immediate` implementation (which just doesn't pass a second argument to `setTimeout`.)

---

After looking harder at that `immediate` implementation, I think the `setTimeout` is actually on a bit of a side path of the control flow -- the thing that's actually at issue here is `process.nextTick`. (Which is a Node thing that's a lot like `setTimeout(…, 0)` but runs before those do.)

But it looks like Jest's fake timers (the "modern" version, which we're using these days) override that too. If I put this in a test:
```js
  const p = new Promise(r => process.nextTick(() => { console.log('B'); r(); }));
  console.log('A');
  //   await p; // this will hang
  jest.advanceTimersByTime(1);
  console.log('C');
  await p; // this is fine
  console.log('D');
```
it prints A, B, C, D in that order. And if I uncomment the `await p` that's before the `advanceTimersByTime`, then it hangs there.

Reply:
See also the comments I just left on https://gist.github.com/apieceofbart/e6dea8d884d29cf88cdb54ef14ddbcc4 , which I found when trying to understand how Jest interacts with `process.nextTick` and had some helpful examples.

Reply:
Cool, this is making more sense.

There's some talk about having "modern" fake timers give you the option to not have `nextTick` mocked: https://github.com/facebook/jest/issues/10221#issuecomment-815803671

But even with a not-mocked `nextTick`, it sounds like we'd still want this change, because `setTimeout` is involved—

```
[…] (The
implementation also uses `setTimeout`, in what seems to be a
convoluted substitute for try/catch.  So that'd be an additional
reason it wouldn't work, if it got that far.)
```

—and Jest's fake timers will always mock `setTimeout`.

Reply:
Yeah.

FWIW now that I understand `process.nextTick` better, I'm fairly confident that the right thing is for Jest to not mock it by default. (Previously I didn't feel I understood the issue well enough to have a firm opinion.)

Details in chat here: https://chat.zulip.org/#narrow/stream/243-mobile-team/topic/Node.20process.2EnextTick/near/1307747
(mainly so that if it ever comes up in the future, we have a good shot at finding the discussion again.)